### ISSUE TITLE
Configurable prices are 0, when MSI is in use and `Show Out Of Stock Products` is enabled.

### ISSUE THREAD

**Issue by dudzio12** [magento/magento2#22000]
### Preconditions (*)
1. Magento 2.3.1
2. Few configurable products
3. Conifgured MSI (ex. 1 source, 1 stock for website)

### Steps to reproduce (*)
1. Create new configurable product.
2. Add custom source (MSI) to all children, remove from them default source
3. Set prices to simples
4. Allow to show out of stock prices on the front.
5. Reindex stock
6. Reindex prices

### Expected result (*)
1. min_price, max_price for configurable should contain correct prices from simple
2. Prices should be visible on front 

### Actual result (*)
1. Prices are 0 for configurable product.

The issue is in here:
\Magento\ConfigurableProduct\Model\ResourceModel\Product\Indexer\Price\Configurable::205

You are joining to the table `cataloginventory_stock_item` which does not contain correct statuses....

Moreover, the if statement is: 
`if ($this->isConfigShowOutOfStock()) { (...) }` 
and when it's true, it's joining to table `cataloginventory_stock_item` with:
`$select->where('si.is_in_stock = ?', Stock::STOCK_IN_STOCK)`, 
which is totally wrong...
The statement should clearly have nagation.

**Comment by magento-engcom-team** [comment]
Hi @dudzio12. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@dudzio12 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by dudzio12** [comment]
@magento-engcom-team give me 2.3-develop instance (2.3.1)

**Comment by magento-engcom-team** [comment]
Hi @dudzio12. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @dudzio12, here is your [Magento instance](https://i-22000-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-22000-2-3-develop.instances.magento-community.engineering/admin](https://i-22000-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
HI @dudzio12 i think you need to report here -> https://github.com/magento-engcom/msi

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by maghamed** [comment]
This is by design behavior in new Inventory (aka MSI). 
The product, in fact, is gone when the order is placed. Thus, it's not possible to place another order with this product, neither to add it to a shopping cart. Even if you will visit the product page - the status of that product would be - "Out of Stock".

The reason why the product still appears on the category page and search result page is that we don't apply Reservations (created as a result of order placement) into consideration on these scenarios. So the product will disappear from these pages as soon as the order would be processed and shipment created. It's quite similar to cache invalidation. The main reason why the system behaves the way it behaves right now is a performance optimization because it's quite a costly operation to retrieve and apply reservations for a bunch of products to render Category and Search result pages.

But along with that, we prevent a possibility of overselling as adding product to the shopping cart we do apply all that checks

**Comment by dudzio12** [comment]
What do you mean - `it's gone`?
So if you buy 1 product, even if stock qty is 100, the product will by unavailable until reservation clears?

Could you tell me what is going on with `cataloginventory_stock_item` `is_in_stock` column? I think it shouldn't be used at all when MSI is enabled. (I agree that this specific issue should be resolved in MSI; price indexer should be rewritten for it).

We have some configurable products, all configured with MSI. I have `inventory_N` table which contains `is_salable` column with final info, if product should be salable or not (in theory, because I don't know if it checks for `min_sale_qty` values, is it?). We need to have `cataloginventory_stock_item` table filled with `is_in_stock` = 1 for everything, otherwise, the products doesn't show up at all. I'm confused in that matter, and I'm not sure if anything should look up to `cataloginventory_stock_item` table, or not. There is no information, anywhere.

I will be glad if you can explain me more technically how MSI workflow works with old inventory tables, because it feels buggy and I'm not sure in which direction I should go to fix them (temporarily).

**Comment by ghost** [comment]
@dudzio12 as this issue related to MSI module i'm will close this one, please  refer to -> https://github.com/magento-engcom/msi/issues/2112

**Comment by maghamed** [comment]
hey @dudzio12 ,
>What do you mean - it's gone?

I mean that Stock availability for that product changed from `In Stock` to `Out of Stock`. It's not possible to buy it or even add to the shopping cart, and even if you visit Product page it would be in the "Out of Stock" status. The only places where product is "In Stock" is on Category Page and on Search Result pages. You can consider that as cache validation. 

>So if you buy 1 product, even if stock qty is 100, the product will by unavailable until reservation clears?

No, a product is getting to "out of stock" when the Salable Quantity reaches 0.
And we use Reservations to prevent overselling. 
You can read more about that here - https://github.com/magento-engcom/msi/wiki/Salable-Quantity-Calculation-and-Mechanism-of-Reservations

>Could you tell me what is going on with cataloginventory_stock_item is_in_stock column? I think it shouldn't be used at all when MSI is enabled. (I agree that this specific issue should be resolved in MSI; price indexer should be rewritten for it).

Both legacy tables (cataloginventory_stock_item/cataloginventory_stock_status) will contain `is_in_stock` = *true* in this case, as these columns don't take reservations into account. And real deduction in these tables will happen when the orders would be shipped. This is the main difference in comparison to how Inventory worked before. 

>I think it shouldn't be used at all when MSI is enabled. 

Correct

>in theory, because I don't know if it checks for min_sale_qty values, is it?

It includes `min_sale_qty`, moreover in MSI we've extended possibility of min_sale_qty + backorder. You can read it here - https://github.com/magento-engcom/msi/wiki/Configure-MSI-backorders

> We need to have cataloginventory_stock_item table filled with is_in_stock = 1 for everything, otherwise

Actually `cataloginventory_stock_item` is used just for `Default Stock`. Thus, it would not affect Custom stocks. 
For each custom stock we create on the fly tables `inventory_stock_N` where N is the stock_id
Check Salability status in these tables for any Stock you are interested in.

> will be glad if you can explain me more technically how MSI workflow works with old inventory tables, because it feels buggy and I'm not sure in which direction I should go to fix them (temporarily).

Now we are using `cataloginventory_stock_item/cataloginventory_stock_status` tables for Default Stock, because we don't have support of Bundle product indexation on custom stocks yet. That's why for Default Stock we use `inventory_stock_1` which in fact is represented as MySQL View but not a real table. This view combines `cataloginventory_stock_status` join on `catalog_product_entity`

**Comment by dudzio12** [comment]
@maghamed - 

Thanks for the reply :)

> Actually cataloginventory_stock_item is used just for Default Stock

It's not true, because even when MSI is used, the configurable product has `is_in_stock` = 1 inside `cataloginventory_stock_item` table (by default). Moreover - stock_item's for configurations (simple products, variants) are created too with `is_in_stock` = 0.

When I change it to 0 or delete that row (configurable stock_item) - configurable product is out of stock and changing qty of variants doesn't affect it at all.

Should i open new issue for that problem?

Tested on vanilla 2.3.1 instance.

**Comment by Thejarnold** [comment]
@dudzio12 
I was looking at this issue, and I believe I have the same exact situation. I am using MSI, I have the system set to show out of stock products, I create a configurable product and remove the products from Default Source and assign the products to a Custom Source. The price on the frontend even with the product being in-stock and having 10,000 units of the product, the category page and product detail page are showing the price being 0.00. If I add the Default Source back to the product alongside the custom source then the price is correctly listed. Did you ever stumble onto a fix for this? 

Thanks for any help or direction you can give.

**Comment by andrewkett** [comment]
We are also seeing the exact same behaviour described by @Thejarnold on 2.3.1. Products are in stock and purchasable as the store is using a custom source, however some of the configurable products are out of stock in cataloginventory_stock_status and are therefore showing $0 on product and category pages

**Comment by tkoszo** [comment]
We have the exact same problem with the out of stock in cataloginventory_stock_status and are therefore showing $0 on the product pages. Is there any fix for this issue?

**Comment by dudzio12** [comment]
I'm interested in it too... @maghamed ?

**Comment by achleitner** [comment]
Bundled product showing price as €0 when out of stock in frontend. Anyone got a fix for this? Thnx
